### DISCUSSION TITLE
Why 'random_steps' function generate random steps with NA in x2_ y2_ and ta_?

### DISCUSSION THREAD

**Discussion by AlineGiroux** [jmsigner/amt#96] [Q&A] [answered]
Me and my colleague @erikagarcez are running iSSA for some species and we just noticed that random_steps function create a butch of NAs in x2_ y2_ and ta_. Can you please explain us why this happen? We have no NA in our original data (x1_ and y1_). I do understand that the first step of a trajectory should have NA in ta_ since there is no a previous direction to calculate the change of direction. I would expect that these NAs in ta_ of real steps could create NAs in ta_ of random steps. Indeed, random steps with NA in ta_ also present NA in x2_ and y2_. So I compared the number of NAs in ta_ to each individual between real and random steps, but they do not correspond. As you can see, some individuals (e.g., berenice) increased their number of NAs from real to random steps, and some of them (e.g., annie) decreased it. 

[na.test.txt](https://github.com/jmsigner/amt/files/13181946/na.test.txt)

`#LOAD DATA AND PREPARE MOVEMENT DATA
resu<-read.table("na.test.txt",sep=",",h=T)
resu$datetime <- as.POSIXct(paste(resu$date.d.m.y.,resu$time),format="%d/%m/%Y %H:%M:%S",tz="Etc/GMT-4")-4*60*60
summary(resu)

#GENERATE RANDOM STEPS USING AMT 
library(amt)
##create a track_xyt object 
tra <- track(resu$x,resu$y,resu$datetime,id_=resu$id,order_by_ts=T,check_duplicates=F,crs=CRS("+proj=utm +zone=21 +south"))

##adjust sampling rate to get regular sampling regime
tra <- tra[order(tra$id_,tra$t_),]
tral <- split(tra,tra$id_)
tral <- lapply(tral,track_resample,rate=minutes(25),tolerance=minutes(15))
for(i in 1:length(tral)){tral[[i]]$burst_=paste(names(tral)[i],tral[[i]]$burst_)}
tra <- do.call("rbind",tral)
summary(tra)

##visual inspection
require(lattice);xyplot(tra$y_~tra$x_|tra$id_)

##convert from points to steps
library(dplyr)
ssf <- tra %>% steps_by_burst(keep_cols="start")
summary(ssf) 
###note: so far we have no NA in x1_ x2_ y1_ or y2_. 
###note: We do have NAs in ta_, as expected in the beggining of the tractories 

##generate random available steps
ssf.r <- ssf %>% random_steps(n=30) 

##check the number of available steps
table(ssf.r$step_id_)
###note: some steps have more than 30 random points. ***Also see burst_***

##investigate if NAs in ta_ correspond to NAs in x2_ and y2_
random <- ssf.r[ssf.r$case_==F,]
table(is.na(random$ta_==T)==is.na(random$x2_==T))
table(is.na(random$ta_==T)==is.na(random$y2_==T))
###note: yes 

##check NAs in ta_ using real steps
table(ssf[is.na(ssf$ta_==T),]$id_)

##check NAs in ta_ using random steps
table(random[is.na(random$ta_==T),]$id_) 
###note: compare 1 and 2 to see how the number of NAs in ta_ do not correspond 

#investigate sl_=0
sl.0 <- ssf.r[ssf.r$sl_==0,]
a <- ssf.r[ssf.r$step_id_==65 & ssf.r$burst_=="annie 107",] #example
###note: why some real steps present sl_=0? 
###note: in the example, we have sl_=0 and ta_=NA on the real step, but no NA in random steps. `

Can you please help us?  Thanks a lot @bsmity13 @jmsigner

**Comment by jmsigner** [comment]
Hi thanks for double-checking this, a few quick answers:

> ###note: some steps have more than 30 random points. Also see burst_

I was not able to find this: 

```
table(table(ssf.r$step_id_))

#    31 
# 192299 
```


> ###note: why some real steps present sl_=0?

It seems you have some 0 sl_ steps in your data, see e.g., 

```
filter(ssf, burst_ == "annie 107", sl_ == 0)
tra |> filter(burst_ == "annie 107", t_ > lubridate::ymd_hms("2015-07-09 00:00:00")) |> slice(17:18)
a <- ssf.r[ssf.r$step_id_==65 & ssf.r$burst_=="annie 107",]
```

**Reply by AlineGiroux** [reply]
Thanks for the quick answers. 
About the random points, when I run my code I get a different result from this. 
![image](https://github.com/jmsigner/amt/assets/91481371/3ed31d38-1b0e-48e6-8910-0d38c46bbe89)

But I noticed that step_ids_ with more than 30 random steps present differences in burst_. Using unique combinations of step_id_ and burst_ we have 30 available steps for each combination. So I believe that this is not a main problem. 

Yes, I have 489 real steps with sl_=0. It's a surprise for me, but in fact this is about my data and not about the package. I thought these real steps with sl_ =0 could generate random steps with NA's in x2_ and y2_. However, it doesn't happen. 
![image](https://github.com/jmsigner/amt/assets/91481371/c20c8836-95f3-49d0-b879-05c4591c865a)

So I still don't understand why I have some random steps with NA's in ta_, x2_ and y2_. 
Can you please help me to understand that? 
thanks a lot

**Comment by bsmity13** [comment, marked as answer]
Hi @AlineGiroux --

As I think you and @jmsigner agree, once you look at the combination of `step_id_` and `burst_`, all strata have exactly 31 steps.

I agree that your issue with `sl_ == 0` is a data issue more than a code issue. You might want to have a look at the cleaning functions in `amt` to help you with those. I have a short tutorial on that available here: https://github.com/bsmity13/data_cleaning. I would point out that those zero-length steps will result in *both* a turn angle of 0 *and* an absolute angle of 0. I think this is actually what's causing the problem you're asking about.

For example, let's look at burst `'annie 107'`.

```
annie107 <- ssf %>% 
  filter(burst_ == "annie 107")

ex1 <- annie107 %>% 
  random_steps(n = 30)
```

In the object `ex1`, step 43 (`step_id_ == 43`) has a step length of 0, so the turn angle is `NA`. Although all the random steps have a turn angle (because they were sampled from the tentative distribution), the entire stratum should be thrown out when you fit the conditional logistic regression model because that stratum will effectively have no observed step. I'm not exactly sure on the behavior of `survival::clogit()` here, but `na.omit(ssf.r)` would remove the observed step but leave the random steps for step 43, and I think that would be problematic for your model.

However, the problem you're actually asking about occurs on step 44. 

```
ex1 %>%
  filter(step_id_ == 44)
```
This is because not only does step 43 not have a turn angle, it does not have a direction at all (there's an NA in the `direction_p` column). So you get additional `NA` values in step 44 because the sum of the angles in 44 and 43 cannot be computed. I.e., although a turn angle can still be sampled (from the tentative distribution), computing `x2_` and `y2_` involves adding the sampled turn angle to the previous absolute angle to get the new absolute angle, and therefore the `NA` in the previous `direction_p` creates `NA`s in `x2_` and `y2_`.

Compare the number of `NA`s if you first exclude steps of length 0.

```
ex2 <- annie107 %>% 
  filter(sl_ > 0) %>% 
  random_steps(n = 30)

# With 0 sl_
sum(is.na(ex1$ta_)) # 32 NAs

# Without 0 sl_
sum(is.na(ex2$ta_)) # 1 NA
```

I think those 0s are the cause of your problem. Take a look and let us know if you agree!

**Comment by AlineGiroux** [comment]
Thanks a lot @bsmity13 and @jmsigner 
Brian, thanks for sharing the cleaning tutorial. Once I removed steps with sl_=0, I didn't find any NA in random steps. So, that was the problem. Thank you.
